File: July-LeetCoding-Challenge/08-3Sum/3Sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums) {
        Array.Sort(nums);
        var result = new List<IList<int>>();
        for(int i=0; i<nums.Length; i++){
            if(i>0 && nums[i] == nums[i-1])
                continue;
            
            int j=i+1, k=nums.Length-1;
            int target = -nums[i];
            while(j < k){
                if(nums[j] + nums[k] == target){
                    result.Add(new int[3]{nums[i], nums[j], nums[k]}.ToList());
                    j++;
                    k--;
                    while(j<k && nums[j] == nums[j-1]) j++;
                    while(j<k && nums[k] == nums[k+1]) k--;
                }else if(nums[j] + nums[k] > target){
                    k--;
                }else if(nums[j] + nums[k] < target){
                    j++;
                }
            }
        }
        return result;        
    }
}

File: July-LeetCoding-Challenge/08-3Sum/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;

        set<vector<int>>s;
        int n = nums.size();
        for (int i = 0; i < n - 2; i++) {
			int left = i + 1, right = n - 1;  
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0)
                    s.insert({nums[i],nums[left++], nums[right--]}); 
                else if (sum < 0) 
                    left++;
                else
                    right--;
            }
        }
	
        for(auto v: s){
            ans.push_back(v);
        }
        return ans;
    }
};